flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsFileDataManager.java

```diff
@@ -192,19 +192,22 @@ public void releaseSubpartitionReader(HsSubpartitionFileReader subpartitionFileR
 
     /** Releases this file data manager and delete shuffle data after all readers is removed. */
     public void release() {
+        List<HsSubpartitionFileReader> pendingReaders;
         synchronized (lock) {
             if (isReleased) {
                 return;
             }
             isReleased = true;
 
-            List<HsSubpartitionFileReader> pendingReaders = new ArrayList<>(allReaders);
+            pendingReaders = new ArrayList<>(allReaders);
             mayNotifyReleased();
+            // delete the shuffle file only when no reader is reading now.
+            releaseFuture.thenRun(this::deleteShuffleFile);
+        }
+        if (!pendingReaders.isEmpty()) {
```

Comment:
I'm not sure about calling `failSubpartitionReaders` outside the lock-guarded scope. The problem is that, readers are not removed from `allReaders` until `failSubpartitionReaders` is called.

I don't see how this change is necessary.

Reply:
done.